plugins/outputs/influxdb_v2/http.go

```diff
@@ -0,0 +1,285 @@
+	ContentEncoding string
+	TLSConfig       *tls.Config
+
+	Serializer *influx.Serializer
+}
+
+type httpClient struct {
+	WriteURL        string
+	ContentEncoding string
+	Timeout         time.Duration
+	Headers         map[string]string
+
+	client     *http.Client
+	serializer *influx.Serializer
+	url        *url.URL
+}
+
+func NewHTTPClient(config *HTTPConfig) (*httpClient, error) {
+	if config.URL == nil {
+		return nil, ErrMissingURL
+	}
+
+	timeout := config.Timeout
+	if timeout == 0 {
+		timeout = defaultRequestTimeout
+	}
+
+	userAgent := config.UserAgent
+	if userAgent == "" {
+		userAgent = defaultUserAgent
+	}
+
+	var headers = make(map[string]string, len(config.Headers)+2)
+	headers["User-Agent"] = userAgent
+	headers["Authorization"] = "Token " + config.Token
+	for k, v := range config.Headers {
+		headers[k] = v
+	}
+
+	var proxy func(*http.Request) (*url.URL, error)
+	if config.Proxy != nil {
+		proxy = http.ProxyURL(config.Proxy)
+	} else {
+		proxy = http.ProxyFromEnvironment
+	}
+
+	serializer := config.Serializer
+	if serializer == nil {
+		serializer = influx.NewSerializer()
+	}
+
+	writeURL, err := makeWriteURL(
+		*config.URL,
+		config.Organization,
+		config.Bucket,
+		config.Precision)
+	if err != nil {
+		return nil, err
+	}
+
+	var transport *http.Transport
+	switch config.URL.Scheme {
+	case "http", "https":
+		transport = &http.Transport{
+			Proxy:           proxy,
+			TLSClientConfig: config.TLSConfig,
+		}
+	case "unix":
+		transport = &http.Transport{
+			Dial: func(_, _ string) (net.Conn, error) {
+				return net.DialTimeout(
+					config.URL.Scheme,
+					config.URL.Path,
+					timeout,
+				)
+			},
+		}
+	default:
+		return nil, fmt.Errorf("unsupported scheme %q", config.URL.Scheme)
+	}
+
+	client := &httpClient{
+		serializer: serializer,
+		client: &http.Client{
+			Timeout:   timeout,
+			Transport: transport,
+		},
+		url:             config.URL,
+		WriteURL:        writeURL,
+		ContentEncoding: config.ContentEncoding,
+		Timeout:         timeout,
+		Headers:         headers,
+	}
+	return client, nil
+}
+
+// URL returns the origin URL that this client connects too.
+func (c *httpClient) URL() string {
+	return c.url.String()
+}
+
+type genericRespError struct {
+	Code      string
+	Message   string
+	Op        string
+	Err       string
+	Line      int32
+	MaxLength int32
+}
+
+func (g genericRespError) String() string {
+	return fmt.Sprintf("%s: %s", g.Code, g.Message)
+}
+
+func (c *httpClient) Write(ctx context.Context, metrics []telegraf.Metric) error {
+	var err error
+
+	reader := influx.NewReader(metrics, c.serializer)
+	req, err := c.makeWriteRequest(reader)
+	if err != nil {
+		return err
+	}
+
+	resp, err := c.client.Do(req.WithContext(ctx))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusNoContent {
+		return nil
+	}
+
+	writeResp := &genericRespError{}
+	json.NewDecoder(resp.Body).Decode(writeResp)
+	var desc string
+
+	switch resp.StatusCode {
```

Comment:
We need to return and error if we would like to retry, and nil if we want to treat the data as written (and delete it).  I think there are three cases:
- Client error and there is potential for a retry to work; return an error
- Client error and no chance a retry will help: log and return nil
- Server error; return an error